java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -0,0 +1,228 @@
+            PrimitiveType.values().forEach {
+                val content = ContentValues().apply {
+                    put("id", it.ordinal)
+                    put("name", it.name)
+                    put("is_primitive", true)
+                }
+                db.insert("types", null, content)
+            }
+        }
+    }
+
+    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) = Unit
+
+    override suspend fun delete(storageKey: StorageKey) {
+        TODO("not implemented")
+    }
+
+    override suspend fun insertOrUpdate(storageKey: StorageKey, entity: Entity) {
+        TODO("not implemented")
+    }
+
+    /** Resets the registrations by deleting everything from the database. */
+    fun reset() {
+        writableDatabase.useTransaction {
+            execSQL("DELETE FROM collection_entries")
+            execSQL("DELETE FROM collections")
+            execSQL("DELETE FROM entities")
+            execSQL("DELETE FROM field_values")
+            execSQL("DELETE FROM fields")
+            execSQL("DELETE FROM number_primitive_values")
+            execSQL("DELETE FROM requested_notifiers")
+            execSQL("DELETE FROM storage_keys")
+            execSQL("DELETE FROM text_primitive_values")
+            execSQL("DELETE FROM types")
+        }
+    }
+
+    fun getOrCreateSchemaTypeId(schema: Schema): TypeId {
+        schemaTypeMap.value[schema.hash]?.let { return it }
+
+        return writableDatabase.useTransaction {
+            val content = ContentValues().apply {
+                put("name", schema.hash)
+                put("is_primitive", false)
+            }
+            val schemaTypeId = insert("types", null, content)
+
+            schemaTypeMap.value[schema.hash] = schemaTypeId
+
+            val insertFieldStatement = compileStatement(
+                "INSERT INTO fields (type_id, parent_type_id, name) VALUES (?, ?, ?)"
+            )
+            val insertFieldBlock = { fieldName: String, fieldType: FieldType ->
+                insertFieldStatement.apply {
+                    bindLong(1, getTypeId(fieldType))
+                    bindLong(2, schemaTypeId)
+                    bindString(3, fieldName)
+                    executeInsert()
+                }
+                Unit
+            }
+            schema.fields.singletons.forEach(insertFieldBlock)
+            schema.fields.collections.forEach(insertFieldBlock)
+            schemaTypeId
+        }
+    }
+
+    /** Returns the type ID for the given [fieldType] if known, otherwise throws. */
+    @VisibleForTesting
+    fun getTypeId(fieldType: FieldType): TypeId = when (fieldType) {
+        is FieldType.Primitive -> fieldType.primitiveType.ordinal.toLong()
+        is FieldType.EntityRef -> schemaTypeMap.value[fieldType.schemaHash]
+            ?: throw IllegalArgumentException(
+                "Unknown type ID for schema with hash ${fieldType.schemaHash}"
+            )
```

Comment:
With `requireNotNull`, this could be simplified to:

```kotlin
is FieldType.EntityRef -> requireNotNull(schemaTypeMap.value[fieldType.schemaHash]) {
    "Unknown type ID for schema with hash ${fieldType.schemaHash}"
}
```

Note: `requireNotNull` throws IllegalArgumentException if null is provided. (Same as with `require`, when its condition is not met)

Reply:
Thanks